conans/client/remote_manager.py

```diff
@@ -29,18 +29,35 @@ def __init__(self, client_cache, remote_client, output):
 
     def upload_conan(self, conan_reference, remote, retry, retry_wait):
         """Will upload the conans to the first remote"""
+
         t1 = time.time()
         export_folder = self._client_cache.export(conan_reference)
+        # make sure that the recipe is complete, with the sources
+        self.get_recipe_sources(conan_reference, export_folder, remote)
         rel_files = relative_dirs(export_folder)
         the_files = {filename: os.path.join(export_folder, filename) for filename in rel_files}
 
         if CONANFILE not in rel_files or CONAN_MANIFEST not in rel_files:
             raise ConanException("Cannot upload corrupted recipe '%s'" % str(conan_reference))
 
+        sources_folder = os.path.join(export_folder, EXPORT_SOURCES_DIR)
+        if os.path.exists(sources_folder):
+            source_files = {k: v for k, v in the_files.items() if k.startswith(EXPORT_SOURCES_DIR)}
+            the_files = {k: v for k, v in the_files.items() if not k.startswith(EXPORT_SOURCES_DIR)}
+
         # FIXME: Check modified exports by hand?
         the_files = compress_conan_files(the_files, export_folder, EXPORT_TGZ_NAME,
                                          CONANFILE, self._output)
 
+        if os.path.exists(sources_folder):
+            sources_tgz = source_files.get(EXPORT_SOURCES_TGZ_NAME)
+            if not sources_tgz:
+                # zip
+                new_files = compress_files(source_files, EXPORT_SOURCES_TGZ_NAME, excluded=[],
+                                           dest_dir=sources_folder)
+                sources_tgz = new_files[EXPORT_SOURCES_TGZ_NAME]
+            the_files[EXPORT_SOURCES_TGZ_NAME] = sources_tgz
+
         ret = self._call_remote(remote, "upload_conan", conan_reference, the_files,
```

Comment:
Add an optional parameter to rest_client.upload_conan called ignore_deleted with a list of files to be ignored if they exist in the remote but not locally. Probably need to add that parameter to auth_manager and the testing equivalent collaborators.